installation_guide/intro_installation.rst: improve the doc

Andersson007 opened the pull request:
installation_guide/intro_installation.rst: improve the doc:
- improve readability by splitting up stuff
- remove not working `See :ref:network_modules` ref
- put `ansible-core` installation example before `ansible` example.
- remove pipX examples. left only mentioning it. i think users can figure it out themselves w/o having the guide overwhelmed with it

Andersson007 committed:
installation_guide/intro_installation.rst: improve the doc

Andersson007 committed:
Fix

Andersson007 committed:
Fix

Andersson007 committed:
Fix

Andersson007 commented:
ready for review

Andersson007 committed:
Remove pipx examples as extra

felixfontein commented:
> remove pipX examples. left only mentioning it. i think users can figure it out themselves w/o having the guide overwhelmed with it

They have only been added recently. Also installing ansible (as opposed to ansible-core) with pipx is more tricky, since you need the ansible-core binaries exposed as well (`--include-deps` flag).

gotmax23 commented:
> remove pipX examples. left only mentioning it. i think users can figure it out themselves w/o having the guide overwhelmed with it

I don't think that's a good idea. `pip install --user` can lead to broken systems, and popular distributions such as Debian and Ubuntu are adopting [PEP 668](https://peps.python.org/pep-0668/) to block users from doing it altogether.  In any case, if you want to remove other people's work, I would first open an issue and invite them to the discussion.

FTR, @sivel added this in https://github.com/ansible/ansible/pull/80833. It seems whether `pip install` or `pipx install` stays at the top [is subject to discussion](https://github.com/ansible/ansible/pull/80833#discussion_r1198267218), but I don't think we should remove `pipx install` entirely.

gotmax23 commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -43,71 +46,28 @@ You can find details about control and managed node requirements for each Ansibl
+The following instructions use ``ansible`` as a package name, but you can substitute ``ansible-core`` if you prefer to start with the minimal package and separately install only the Ansible Collections you require.
 
-.. _pipx_upgrade:
+The ``ansible`` or ``ansible-core`` packages may be available in your operating systems package manager, and you are free to install these packages with your preferred method. These installation instructions only cover the officially supported means of installing the python package with ``pip``.
```
Should this link to the installation_distros doc?

Andersson007 replied:
IMO good idea, added, thanks! PTAL

Andersson007 committed:
Return pipx

Andersson007 commented:
> > remove pipX examples. left only mentioning it. i think users can figure it out themselves w/o having the guide overwhelmed with it
> 
> I don't think that's a good idea. `pip install --user` can lead to broken systems, and popular distributions such as Debian and Ubuntu are adopting [PEP 668](https://peps.python.org/pep-0668/) to block users from doing it altogether. In any case, if you want to remove other people's work, I would first open an issue and invite them to the discussion.
> 
> FTR, @sivel added this in [ansible/ansible#80833](https://github.com/ansible/ansible/pull/80833). It seems whether `pip install` or `pipx install` stays at the top [is subject to discussion](https://github.com/ansible/ansible/pull/80833#discussion_r1198267218), but I don't think we should remove `pipx install` entirely.

@gotmax23 @felixfontein i've put it back, thanks. Though the order now is that pip will go first.

Andersson007 committed:
Add ref to installation_distros guide

Andersson007 committed:
Remove some extra spaces

Andersson007 commented:
ready for review

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -22,12 +24,12 @@ For your *control* node (the machine that runs Ansible), you can use nearly any
 .. note::
 
-   Network modules are an exception and do not require Python on the managed device.  See :ref:`network_modules`.
+   There can be exceptions in module requirements. For example, network modules do not require Python on the managed device. See documentation for modules you use.
```
```suggestion
   There can be exceptions in module requirements. For example, network modules do not require Python on the managed device. See documentation for the modules you use.
```

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -43,71 +45,26 @@ You can find details about control and managed node requirements for each Ansibl
-.. code-block:: console
-
-    $ pipx install ansible-core
+* ``ansible-core``: a minimalist language and runtime package containing a set of `builtin modules and plugins <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/index.html>`_.
```
This should use a :ref:  I think :ref:`plugins_in_ansible.builtin` should work for both core and package docs.

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -143,18 +101,24 @@ You may need to perform some additional configuration before you are able to run
 ------------------
 
-Use ``pip`` in your selected Python environment to install the Ansible package of your choice for the current user:
+Use ``pip`` in your selected Python environment to install the minimal ``ansible-core`` package for the current user:
```
Changing the order here changes what this heading is.  The original had Ansible package first as that's the heading "Installing Ansible".  This was a conscious decision to make Ansible the package the prominent and preferred installation.

If community folks feel differently, then this needs to be two separate sections now with a heading for core and a heading for package installs.

Andersson007 replied:
Changed the order as it was

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -143,18 +101,24 @@ You may need to perform some additional configuration before you are able to run
+    $ python3 -m pip install --user ansible-core
 
-Alternately, you can install a specific version of ``ansible-core`` in this Python environment:
+You can install a specific version of ``ansible-core`` if you need:
```
not sure I agree that this change makes it more readable. 'if you need' is an odd expression to end the sentence with. The original "Alternately" clearly indicates this is optional and an alternate to the prior install step.

Andersson007 replied:
Changed as it was, thanks for explaining

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -143,18 +101,24 @@ You may need to perform some additional configuration before you are able to run
 
     $ python3 -m pip install --user ansible-core==2.12.3
 
+Alternatively, you can install the full "batteries-included" Ansible package for the current user:
```
just to repeat - either Ansible package stays first, or this is split into two sections one for core, one for package.

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -168,6 +132,60 @@ To upgrade an existing Ansible installation in this Python environment to the la
+
+    $ pipx install ansible-core==2.12.3
+
+Alternatively, you can install the full Ansible package:
```
Same comments apply. The original text was written in that order for reasons stated above. Either keep the original wording and order or separate into two sections.  S/Alternatively/Alternately/

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -262,24 +282,24 @@ To check the version of the ``ansible`` package that has been installed:
 =======================================
 
-You can add shell completion of the Ansible command line utilities by installing an optional dependency called ``argcomplete``. ``argcomplete`` supports bash, and has limited support for zsh and tcsh.
+You can add shell completion of the Ansible command line utilities by installing an optional dependency called ``argcomplete``, which supports bash, and has limited support for zsh and tcsh.
```
This change makes the sentence less readable. Longer sentences are harder to understand, especially by non-native English speakers. The original is two sentences for this reason.

Andersson007 replied:
I agree. The reason i made it was that the repetition (esp. as it's lowercase) didn't look good. I've replaced it with just `it`. PTAL

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -262,24 +282,24 @@ To check the version of the ``ansible`` package that has been installed:
 --------------------------
 
-If you chose the ``pipx`` install instructions:
+If you chose the ``pip``:
```
original sentence is clearer.

samccann commented on docs/docsite/rst/installation_guide/intro_installation.rst:
```diff
@@ -262,24 +282,24 @@ To check the version of the ``ansible`` package that has been installed:
+    $ python3 -m pip install --user argcomplete
 
-Or, if you chose the ``pip`` install instructions:
+If you chose the ``pipx``:
```
Original sentence is clearer, though you can remove the 'Or,`

Andersson007 committed:
Update docs/docsite/rst/installation_guide/intro_installation.rst

Co-authored-by: Sandra McCann <samccann@redhat.com>

Andersson007 committed:
argcomplete

Andersson007 committed:
Change ref to builtin modules and plugins

Andersson007 committed:
Incorporate feedback

Andersson007 committed:
Incorporate feedback

Andersson007 committed:
Fix formatting

Andersson007 commented:
@samccann 
- Changed the order as it was: ansible first, then ansible-core
- Changed the wording using `alternately`

Thanks for the feedback, PTAL

gotmax23 commented:
I requested my own review. I'd like to look at this again before it's merged.

gotmax23 commented:
I think the pipx instructions should stay above the pip instructions. pipx is a safer way to install Python applications and is more widely compatible.

samccann commented:
We also need approval from someone like @sivel from the core side.

sivel commented:
I suppose I could go either way on pip vs pipx being at the top, but I'll explain why I put it at the top originally.  With PEP668 becoming more prevalent, see https://github.com/ansible/ansible-documentation/issues/100, we had debated whether we should go down the path of recommending using a virtualenv.  Ultimately we decided management of a virtualenv was overly complicated for the average user.

Due to this, we decided that the pipx instructions will better serve the majority of users as more distros adopt PEP668.

Andersson007 committed:
move pipx paragraph back

Andersson007 committed:
fix

Andersson007 committed:
fix

Andersson007 commented:
@gotmax23 @sivel thanks for providing the context! I've put it back and TIL, thanks

@samccann i think the PR could be merged, PTAL

Andersson007 commented:
Thanks everyone for the great reviews!:)
